File: typescript/tests/parser.test.mjs
import test from 'node:test'
import assert from 'node:assert/strict'
import { writeFileSync, unlinkSync } from 'node:fs'

import { parseArgs, resolveCommand, RUNNER_REGISTRY } from '../src/parser.js'
import { loadConfig } from '../src/config.js'

test('parseArgs: prompt-only', () => {
  const parsed = parseArgs(['fix the bug'])
  assert.deepEqual(parsed, {
    runner: null,
    thinking: null,
    provider: null,
    model: null,
    alias: null,
    prompt: 'fix the bug',
  })
})

test('parseArgs: runner selector', () => {
  const parsed = parseArgs(['claude', 'do the thing'])
  assert.equal(parsed.runner, 'claude')
  assert.equal(parsed.prompt, 'do the thing')
})

test('parseArgs: runner abbreviation cc', () => {
  const parsed = parseArgs(['cc', 'prompt'])
  assert.equal(parsed.runner, 'cc')
})

test('parseArgs: runner abbreviation oc', () => {
  const parsed = parseArgs(['oc', 'prompt'])
  assert.equal(parsed.runner, 'oc')
})

test('parseArgs: runner abbreviation k', () => {
  const parsed = parseArgs(['k', 'prompt'])
  assert.equal(parsed.runner, 'k')
})

test('parseArgs: runner abbreviation rc', () => {
  const parsed = parseArgs(['rc', 'prompt'])
  assert.equal(parsed.runner, 'rc')
})

test('parseArgs: runner abbreviation cr', () => {
  const parsed = parseArgs(['cr', 'prompt'])
  assert.equal(parsed.runner, 'cr')
})

test('parseArgs: thinking flag', () => {
  const parsed = parseArgs(['+3', 'prompt'])
  assert.equal(parsed.thinking, 3)
  assert.equal(parsed.prompt, 'prompt')
})

test('parseArgs: thinking flag zero', () => {
  const parsed = parseArgs(['+0', 'prompt'])
  assert.equal(parsed.thinking, 0)
})

test('parseArgs: provider:model', () => {
  const parsed = parseArgs([':openai:gpt-4o', 'prompt'])
  assert.equal(parsed.provider, 'openai')
  assert.equal(parsed.model, 'gpt-4o')
})

test('parseArgs: model only', () => {
  const parsed = parseArgs([':gpt-4o', 'prompt'])
  assert.equal(parsed.model, 'gpt-4o')
  assert.equal(parsed.provider, null)
})

test('parseArgs: alias', () => {
  const parsed = parseArgs(['@work', 'prompt'])
  assert.equal(parsed.alias, 'work')
})

test('parseArgs: full combo', () => {
  const parsed = parseArgs(['claude', '+2', ':anthropic:sonnet', '@fast', 'fix it'])
  assert.equal(parsed.runner, 'claude')
  assert.equal(parsed.thinking, 2)
  assert.equal(parsed.provider, 'anthropic')
  assert.equal(parsed.model, 'sonnet')
  assert.equal(parsed.alias, 'fast')
  assert.equal(parsed.prompt, 'fix it')
})

test('parseArgs: multi-word prompt', () => {
  const parsed = parseArgs(['fix', 'all', 'the', 'bugs'])
  assert.equal(parsed.prompt, 'fix all the bugs')
})

test('parseArgs: runner after positional becomes positional', () => {
  const parsed = parseArgs(['hello', 'claude'])
  assert.equal(parsed.runner, null)
  assert.equal(parsed.prompt, 'hello claude')
})

test('parseArgs: alias after positional is positional', () => {
  const parsed = parseArgs(['hello', '@work'])
  assert.equal(parsed.alias, null)
  assert.equal(parsed.prompt, 'hello @work')
})

test('resolveCommand: default runner (opencode)', () => {
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: 'do stuff' })
  assert.deepEqual(result.argv[0], 'opencode')
  assert.equal(result.argv[result.argv.length - 1], 'do stuff')
})

test('resolveCommand: claude runner', () => {
  const result = resolveCommand({ runner: 'claude', thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' })
  assert.equal(result.argv[0], 'claude')
})

test('resolveCommand: thinking flags for claude +2', () => {
  const result = resolveCommand({ runner: 'claude', thinking: 2, provider: null, model: null, alias: null, prompt: 'prompt' })
  assert.deepEqual(result.argv.slice(1, 3), ['--thinking', 'medium'])
})

test('resolveCommand: thinking flags for claude +0', () => {
  const result = resolveCommand({ runner: 'claude', thinking: 0, provider: null, model: null, alias: null, prompt: 'prompt' })
  assert.deepEqual(result.argv.slice(1, 2), ['--no-thinking'])
})

test('resolveCommand: thinking flags for kimi +3', () => {
  const result = resolveCommand({ runner: 'kimi', thinking: 3, provider: null, model: null, alias: null, prompt: 'prompt' })
  assert.deepEqual(result.argv.slice(1, 3), ['--think', 'high'])
})

test('resolveCommand: model flag for claude', () => {
  const result = resolveCommand({ runner: 'claude', thinking: null, provider: null, model: 'sonnet', alias: null, prompt: 'prompt' })
  assert.ok(result.argv.includes('--model'))
  assert.ok(result.argv.includes('sonnet'))
})

test('resolveCommand: provider env override', () => {
  const result = resolveCommand({ runner: null, thinking: null, provider: 'openai', model: null, alias: null, prompt: 'prompt' })
  assert.equal(result.env.CCC_PROVIDER, 'openai')
})

test('resolveCommand: no provider env when empty', () => {
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' })
  assert.deepEqual(result.env, {})
})

test('resolveCommand: empty prompt throws', () => {
  assert.throws(
    () => resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: '   ' }),
    { message: 'prompt must not be empty' },
  )
})

test('resolveCommand: config default runner', () => {
  const config = {
    defaultRunner: 'claude',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: null,
    aliases: {},
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' }, config)
  assert.equal(result.argv[0], 'claude')
})

test('resolveCommand: config default thinking', () => {
  const config = {
    defaultRunner: 'claude',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: 1,
    aliases: {},
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' }, config)
  assert.ok(result.argv.includes('--thinking'))
  assert.ok(result.argv.includes('low'))
})

test('resolveCommand: config default model', () => {
  const config = {
    defaultRunner: 'claude',
    defaultProvider: '',
    defaultModel: 'sonnet',
    defaultThinking: null,
    aliases: {},
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' }, config)
  assert.ok(result.argv.includes('--model'))
  assert.ok(result.argv.includes('sonnet'))
})

test('resolveCommand: config default provider', () => {
  const config = {
    defaultRunner: 'oc',
    defaultProvider: 'openai',
    defaultModel: '',
    defaultThinking: null,
    aliases: {},
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' }, config)
  assert.equal(result.env.CCC_PROVIDER, 'openai')
})

test('resolveCommand: alias resolution', () => {
  const config = {
    defaultRunner: 'oc',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: null,
    aliases: {
      fast: { runner: 'claude', model: 'sonnet' },
    },
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: 'fast', prompt: 'prompt' }, config)
  assert.equal(result.argv[0], 'claude')
  assert.ok(result.argv.includes('sonnet'))
  assert.deepEqual(result.warnings, [])
})

test('resolveCommand: alias thinking override', () => {
  const config = {
    defaultRunner: 'claude',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: null,
    aliases: {
      deep: { thinking: 4 },
    },
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: 'deep', prompt: 'prompt' }, config)
  assert.ok(result.argv.includes('--thinking'))
  assert.ok(result.argv.includes('max'))
  assert.deepEqual(result.warnings, [])
})

test('resolveCommand: alias fallback uses agent when preset missing', () => {
  const result = resolveCommand(
    { runner: null, thinking: null, provider: null, model: null, alias: 'reviewer', prompt: 'prompt' },
    {
      defaultRunner: 'oc',
      defaultProvider: '',
      defaultModel: '',
      defaultThinking: null,
      aliases: {},
      abbreviations: {},
    },
  )
  assert.deepEqual(result.argv.slice(0, 4), ['opencode', 'run', '--agent', 'reviewer'])
  assert.deepEqual(result.warnings, [])
})

test('resolveCommand: preset agent is emitted when configured', () => {
  const config = {
    defaultRunner: 'oc',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: null,
    aliases: {
      review: { agent: 'specialist' },
    },
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: null, provider: null, model: null, alias: 'review', prompt: 'prompt' }, config)
  assert.deepEqual(result.argv.slice(0, 4), ['opencode', 'run', '--agent', 'specialist'])
  assert.deepEqual(result.warnings, [])
})

test('resolveCommand: unsupported agent emits warning and no agent flag', () => {
  const result = resolveCommand(
    { runner: 'rc', thinking: null, provider: null, model: null, alias: 'reviewer', prompt: 'prompt' },
    {
      defaultRunner: 'oc',
      defaultProvider: '',
      defaultModel: '',
      defaultThinking: null,
      aliases: {},
      abbreviations: {},
    },
  )
  assert.equal(result.argv[0], 'codex')
  assert.ok(!result.argv.includes('--agent'))
  assert.deepEqual(result.warnings, ['warning: runner "rc" does not support agents; ignoring @reviewer'])
})

test('resolveCommand: abbreviation resolution', () => {
  const config = {
    defaultRunner: 'oc',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: null,
    aliases: {},
    abbreviations: { myrunner: 'claude' },
  }
  const result = resolveCommand({ runner: 'myrunner', thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' }, config)
  assert.equal(result.argv[0], 'claude')
})

test('resolveCommand: unknown runner falls back to default', () => {
  const result = resolveCommand({ runner: 'unknown', thinking: null, provider: null, model: null, alias: null, prompt: 'prompt' })
  assert.equal(result.argv[0], 'opencode')
})

test('resolveCommand: parsed thinking overrides alias thinking', () => {
  const config = {
    defaultRunner: 'claude',
    defaultProvider: '',
    defaultModel: '',
    defaultThinking: null,
    aliases: {
      deep: { thinking: 4 },
    },
    abbreviations: {},
  }
  const result = resolveCommand({ runner: null, thinking: 1, provider: null, model: null, alias: 'deep', prompt: 'prompt' }, config)
  assert.ok(result.argv.includes('low'))
})

test('RUNNER_REGISTRY: has all expected runners', () => {
  assert.ok(RUNNER_REGISTRY.opencode)
  assert.ok(RUNNER_REGISTRY.claude)
  assert.ok(RUNNER_REGISTRY.kimi)
  assert.ok(RUNNER_REGISTRY.codex)
  assert.ok(RUNNER_REGISTRY.crush)
})

test('RUNNER_REGISTRY: agent flags for supported runners', () => {
  assert.equal(RUNNER_REGISTRY.opencode.agentFlag, '--agent')
  assert.equal(RUNNER_REGISTRY.claude.agentFlag, '--agent')
  assert.equal(RUNNER_REGISTRY.kimi.agentFlag, '--agent')
  assert.equal(RUNNER_REGISTRY.codex.agentFlag, '')
  assert.equal(RUNNER_REGISTRY.crush.agentFlag, '')
})

test('RUNNER_REGISTRY: abbreviations point to same objects', () => {
  assert.equal(RUNNER_REGISTRY.oc, RUNNER_REGISTRY.opencode)
  assert.equal(RUNNER_REGISTRY.cc, RUNNER_REGISTRY.claude)
  assert.equal(RUNNER_REGISTRY.c, RUNNER_REGISTRY.claude)
  assert.equal(RUNNER_REGISTRY.k, RUNNER_REGISTRY.kimi)
  assert.equal(RUNNER_REGISTRY.rc, RUNNER_REGISTRY.codex)
  assert.equal(RUNNER_REGISTRY.cr, RUNNER_REGISTRY.crush)
})

test('loadConfig: returns defaults when no file', () => {
  const config = loadConfig('/nonexistent/path/config.toml')
  assert.equal(config.defaultRunner, 'oc')
  assert.equal(config.defaultProvider, '')
  assert.equal(config.defaultModel, '')
  assert.equal(config.defaultThinking, null)
  assert.deepEqual(config.aliases, {})
  assert.deepEqual(config.abbreviations, {})
})

test('loadConfig: parses agent in alias presets', () => {
  const configToml = `
[aliases.review]
runner = "claude"
thinking = 2
  provider = "anthropic"
  model = "sonnet"
  agent = "specialist"
`
  const path = `/tmp/ccc-typescript-agent-config-${process.pid}.toml`
  writeFileSync(path, configToml)
  try {
    const config = loadConfig(path)
    assert.equal(config.aliases.review.agent, 'specialist')
    assert.equal(config.aliases.review.runner, 'claude')
  } finally {
    unlinkSync(path)
  }
})

File: zig/tests/parser_test.zig
const std = @import("std");
const parser = @import("parser");

test "parseArgs prompt-only" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "hello", "world" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.runner == null);
    try std.testing.expect(parsed.thinking == null);
    try std.testing.expect(parsed.provider == null);
    try std.testing.expect(parsed.model == null);
    try std.testing.expect(parsed.alias == null);
    try std.testing.expectEqualStrings("hello world", parsed.prompt);
}

test "parseArgs runner selector" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "claude", "do stuff" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.runner != null);
    try std.testing.expectEqualStrings("claude", parsed.runner.?);
    try std.testing.expectEqualStrings("do stuff", parsed.prompt);
}

test "parseArgs runner selector case insensitive" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "Claude", "do stuff" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.runner != null);
    try std.testing.expectEqualStrings("claude", parsed.runner.?);
}

test "parseArgs runner abbreviation" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "cc", "do stuff" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.runner != null);
    try std.testing.expectEqualStrings("cc", parsed.runner.?);
}

test "parseArgs thinking" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "+3", "think deeply" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.thinking != null);
    try std.testing.expectEqual(@as(i32, 3), parsed.thinking.?);
    try std.testing.expectEqualStrings("think deeply", parsed.prompt);
}

test "parseArgs thinking zero" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "+0", "no think" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expectEqual(@as(i32, 0), parsed.thinking.?);
}

test "parseArgs provider and model" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ ":openai:gpt-4", "prompt" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.provider != null);
    try std.testing.expectEqualStrings("openai", parsed.provider.?);
    try std.testing.expect(parsed.model != null);
    try std.testing.expectEqualStrings("gpt-4", parsed.model.?);
}

test "parseArgs model only" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ ":gpt-4o", "prompt" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.provider == null);
    try std.testing.expect(parsed.model != null);
    try std.testing.expectEqualStrings("gpt-4o", parsed.model.?);
}

test "parseArgs alias" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "@fast", "prompt" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expect(parsed.alias != null);
    try std.testing.expectEqualStrings("fast", parsed.alias.?);
}

test "parseArgs full combo" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "claude", "+2", ":anthropic:claude-3.5-sonnet", "@fast", "write code" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expectEqualStrings("claude", parsed.runner.?);
    try std.testing.expectEqual(@as(i32, 2), parsed.thinking.?);
    try std.testing.expectEqualStrings("anthropic", parsed.provider.?);
    try std.testing.expectEqualStrings("claude-3.5-sonnet", parsed.model.?);
    try std.testing.expectEqualStrings("fast", parsed.alias.?);
    try std.testing.expectEqualStrings("write code", parsed.prompt);
}

test "parseArgs positional after runner eats non-special" {
    const allocator = std.testing.allocator;
    const argv = &[_][]const u8{ "+3", "not+a", "flag" };
    var parsed = try parser.parseArgs(allocator, argv);
    defer parsed.deinit(allocator);

    try std.testing.expectEqual(@as(i32, 3), parsed.thinking.?);
    try std.testing.expectEqualStrings("not+a flag", parsed.prompt);
}

test "resolveCommand default runner" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{"hello"});
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expect(resolved.argv.len >= 3);
    try std.testing.expectEqualStrings("opencode", resolved.argv[0]);
    try std.testing.expectEqualStrings("run", resolved.argv[1]);
    try std.testing.expectEqualStrings("hello", resolved.argv[2]);
}

test "resolveCommand claude runner" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "claude", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expect(resolved.argv.len >= 2);
    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[1]);
}

test "resolveCommand thinking flags claude" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "claude", "+3", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("--thinking", resolved.argv[1]);
    try std.testing.expectEqualStrings("high", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
}

test "resolveCommand thinking flags kimi" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "kimi", "+0", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("kimi", resolved.argv[0]);
    try std.testing.expectEqualStrings("--no-think", resolved.argv[1]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[2]);
}

test "resolveCommand model flag" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "claude", ":gpt-4o", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("--model", resolved.argv[1]);
    try std.testing.expectEqualStrings("gpt-4o", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
}

test "resolveCommand provider env" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ ":openai:gpt-4", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    const provider = resolved.env.get("CCC_PROVIDER");
    try std.testing.expect(provider != null);
    try std.testing.expectEqualStrings("openai", provider.?);
}

test "resolveCommand empty prompt error" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{"claude"});
    defer parsed.deinit(allocator);

    const result = parser.resolveCommand(allocator, parsed, &config);
    try std.testing.expectError(error.EmptyPrompt, result);
}

test "resolveCommand alias runner override" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    try config.aliases.put("fast", .{
        .runner = "claude",
        .thinking = 1,
    });

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "@fast", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("--thinking", resolved.argv[1]);
    try std.testing.expectEqualStrings("low", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
}

test "resolveCommand name falls back to agent" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "@reviewer", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("opencode", resolved.argv[0]);
    try std.testing.expectEqualStrings("run", resolved.argv[1]);
    try std.testing.expectEqualStrings("--agent", resolved.argv[2]);
    try std.testing.expectEqualStrings("reviewer", resolved.argv[3]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[4]);
    try std.testing.expect(resolved.warnings.len == 0);
}

test "resolveCommand preset agent wins over name fallback" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    try config.aliases.put("reviewer", .{
        .agent = "specialist",
    });

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "@reviewer", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("opencode", resolved.argv[0]);
    try std.testing.expectEqualStrings("run", resolved.argv[1]);
    try std.testing.expectEqualStrings("--agent", resolved.argv[2]);
    try std.testing.expectEqualStrings("specialist", resolved.argv[3]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[4]);
    try std.testing.expect(resolved.warnings.len == 0);
}

test "resolveCommand claude agent flag" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "claude", "@reviewer", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("--agent", resolved.argv[1]);
    try std.testing.expectEqualStrings("reviewer", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
    try std.testing.expect(resolved.warnings.len == 0);
}

test "resolveCommand kimi agent flag" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "k", "@reviewer", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("kimi", resolved.argv[0]);
    try std.testing.expectEqualStrings("--agent", resolved.argv[1]);
    try std.testing.expectEqualStrings("reviewer", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
    try std.testing.expect(resolved.warnings.len == 0);
}

test "resolveCommand unsupported agent warns" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "rc", "@reviewer", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("codex", resolved.argv[0]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[1]);
    try std.testing.expect(resolved.warnings.len == 1);
    try std.testing.expectEqualStrings(
        "warning: runner codex does not support agents; ignoring @reviewer",
        resolved.warnings[0],
    );
}

test "resolveCommand default thinking from config" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();
    config.default_thinking = 4;

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "claude", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("--thinking", resolved.argv[1]);
    try std.testing.expectEqualStrings("max", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
}

test "resolveCommand codex with model" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "codex", ":gpt-4", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("codex", resolved.argv[0]);
    try std.testing.expectEqualStrings("--model", resolved.argv[1]);
    try std.testing.expectEqualStrings("gpt-4", resolved.argv[2]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[3]);
}

test "resolveCommand crush no model flag" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "crush", ":model-1", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("crush", resolved.argv[0]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[1]);
    try std.testing.expect(resolved.argv.len == 2);
}

test "resolveCommand abbreviation from config" {
    const allocator = std.testing.allocator;
    var config = parser.CccConfig.init(allocator);
    defer config.deinit();

    try config.abbreviations.put("oc", "claude");

    var parsed = try parser.parseArgs(allocator, &[_][]const u8{ "oc", "prompt" });
    defer parsed.deinit(allocator);

    var resolved = try parser.resolveCommand(allocator, parsed, &config);
    defer resolved.deinit(allocator);

    try std.testing.expectEqualStrings("claude", resolved.argv[0]);
    try std.testing.expectEqualStrings("prompt", resolved.argv[1]);
}

test "runnerRegistry has all entries" {
    const allocator = std.testing.allocator;
    var registry = try parser.runnerRegistry(allocator);
    defer registry.deinit();

    try std.testing.expect(registry.get("opencode") != null);
    try std.testing.expect(registry.get("claude") != null);
    try std.testing.expect(registry.get("kimi") != null);
    try std.testing.expect(registry.get("codex") != null);
    try std.testing.expect(registry.get("crush") != null);
    try std.testing.expect(registry.get("oc") != null);
    try std.testing.expect(registry.get("cc") != null);
    try std.testing.expect(registry.get("c") != null);
    try std.testing.expect(registry.get("k") != null);
    try std.testing.expect(registry.get("rc") != null);
    try std.testing.expect(registry.get("cr") != null);

    const oc = registry.get("oc").?;
    const opencode = registry.get("opencode").?;
    try std.testing.expectEqualStrings(opencode.binary, oc.binary);
}
